### PR TITLE
[cdc-pipeline-connector][mysql] Introduce mysql cdc pipeline DataSource

### LINKED ISSUE #2647: [flink-cdc-pipeline-connectors] Add Implementation of DataSource in MySQL
### Search before asking

- [X] I searched in the [issues](https://github.com/ververica/flink-cdc-connectors/issues) and found nothing similar.

### Motivation

After https://github.com/ververica/flink-cdc-connectors/pull/2638 merged, we can try to add implementation in MySQL to simulate enterprise scenarios.

You may need to wait for https://github.com/ververica/flink-cdc-connectors/issues/2642 https://github.com/ververica/flink-cdc-connectors/issues/2644 completed. 

### Solution

_No response_

### Alternatives

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by Jiabao-Sun**
Introduce mysql cdc pipeline DataSource

Resolve #2647.

### CODE REVIEW

File: flink-cdc-common/src/main/java/com/ververica/cdc/common/event/EventDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ververica.cdc.common.event;
+
+import java.io.Serializable;
+import java.util.Optional;
+
+/** Deserializer to deserialize given record to {@link Event}. */
```
Comment (leonardBang) [inline flink-cdc-common/src/main/java/com/ververica/cdc/common/event/EventDeserializer.java:22]: Could be a public interface here

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/factory/MySqlDataFactory.java [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+import static com.ververica.cdc.connectors.mysql.source.MySqlDataSourceOptions.TABLES;
+import static com.ververica.cdc.connectors.mysql.source.MySqlDataSourceOptions.USERNAME;
+import static com.ververica.cdc.connectors.mysql.source.utils.ObjectUtils.doubleCompare;
+import static com.ververica.cdc.debezium.table.DebeziumOptions.getDebeziumProperties;
+import static com.ververica.cdc.debezium.utils.JdbcUrlUtils.getJdbcProperties;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** A {@link Factory} to create {@link MySqlDataSource}. */
+@Internal
+public class MySqlDataFactory implements DataSourceFactory {
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/factory/MySqlDataFactory.java:82]: MySqlDataSourceFactory ? we may add MySqlDataSinkFactory later

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/factory/MySqlDataFactory.java [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+                        .username(username)
+                        .password(password)
+                        .databaseList(".*")
+                        .tableList(".*")
+                        .startupOptions(startupOptions)
+                        .serverId(serverId)
+                        .serverTimeZone(serverTimeZone.getId())
+                        .fetchSize(fetchSize)
+                        .splitSize(splitSize)
+                        // TODO: .chunkKeyColumn()
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/factory/MySqlDataFactory.java:152]: we can create an issue to support chunKeyColumn in table level in mysql cdc connector

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/factory/MySqlDataFactory.java [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+                return StartupOptions.latest();
+
+            case SCAN_STARTUP_MODE_VALUE_EARLIEST:
+                return StartupOptions.earliest();
+
+            case SCAN_STARTUP_MODE_VALUE_SPECIFIC_OFFSET:
+                validateSpecificOffset(config);
+                return getSpecificOffset(config);
+
+            case SCAN_STARTUP_MODE_VALUE_TIMESTAMP:
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/factory/MySqlDataFactory.java:245]: note: All of these mode need to send create table event firstly

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/source/MySqlDataSourceOptions.java [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+                    .stringType()
+                    .noDefaultValue()
+                    .withDescription(
+                            "Password to use when connecting to the MySQL database server.");
+    public static final ConfigOption<String> TABLES =
+            ConfigOptions.key("tables")
+                    .stringType()
+                    .noDefaultValue()
+                    .withDescription(
+                            "Table names of the MySQL tables to monitor. Regular expressions are supported. eg. adb.*, bdb.user_table_[0-9]+, [app|web]_order_.*");
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/source/MySqlDataSourceOptions.java:60]: could you add escape character explanation in description? and we can also explain the table id mapping relation here, mysql's database name should map to our schemaName, mysql's table name should map to our tableName

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/source/MySqlMetadataAccessor.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public List<String> listSchemas(@Nullable String namespace) {
+        return MySqlSchemaUtils.listDatabases(sourceConfig);
+    }
+
+    @Override
+    public List<TableId> listTables(@Nullable String namespace, @Nullable String dbName) {
+        return MySqlSchemaUtils.listTables(sourceConfig, dbName);
+    }
+
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/source/MySqlMetadataAccessor.java:59]: could you add some java doc for the three methods? the note could contains the table path mapping relation

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/utils/MySqlTypeUtils.java [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+     */
+    private static DataType convertFromColumn(Column column) {
+        String typeName = column.typeName();
+        switch (typeName) {
+            case TINYINT:
+                // MySQL haven't boolean type, it uses tinyint(1) to represents boolean type
+                // user should not use tinyint(1) to store number although jdbc url parameter
+                // tinyInt1isBit=false can help change the return value, it's not a general way
+                // btw: mybatis and mysql-connector-java map tinyint(1) to boolean by default
+                return column.length() == 1 ? DataTypes.BOOLEAN() : DataTypes.TINYINT();
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/utils/MySqlTypeUtils.java:112]: We can introduce a `treat-tinyint1-as-boolean` config for mysql data source, a `TINYINT (1)` type in mysql can store value `2`,`3` too.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/utils/MySqlTypeUtils.java [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+            case DECIMAL_UNSIGNED_ZEROFILL:
+                return column.length() <= 38
+                        ? DataTypes.DECIMAL(column.length(), column.scale().orElse(0))
+                        : DataTypes.STRING();
+            case TIME:
+                return column.length() >= 0 ? DataTypes.TIME(column.length()) : DataTypes.TIME();
+            case DATE:
+                return DataTypes.DATE();
+            case DATETIME:
+            case TIMESTAMP:
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/utils/MySqlTypeUtils.java:163]: TIMESTAMP should map to `TIMESTAMP_LTZ` if IIUC, we map it wrong in mysql cdc connector, it's better if we can change it in our first version.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/resources/META-INF/services/com.ververica.cdc.common.factories.Factory [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/resources/META-INF/services/com.ververica.cdc.common.factories.Factory:1]: minor: Ververica copyright

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/test/resources/pipeline-test.yaml [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+source:
+  type: mysql
+  name: MySQL Source
+  hostname: 172.29.0.237
+  port: 3306
+  username: root
+  password: AAbb0011
+  tables: atta_crm.\.*
+#  chunk-column: app_order_.*:id,web_order:product_id
+#  capture-new-tables: true
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/test/resources/pipeline-test.yaml:24]: We can release this feature in next release version, I'm worried we can not test well under current status

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-debezium/src/main/java/com/ververica/cdc/debezium/event/DebeziumEventDeserializationSchema.java [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+    /** Creates a runtime converter which is null safe. */
+    private static DeserializationRuntimeConverter createConverter(
+            DataType type, ZoneId serverTimeZone) {
+        return wrapIntoNullableConverter(createNotNullConverter(type, serverTimeZone));
+    }
+
+    // --------------------------------------------------------------------------------
+    // IMPORTANT! We use anonymous classes instead of lambdas for a reason here. It is
+    // necessary because the maven shade plugin cannot relocate classes in
+    // SerializedLambdas (MSHADE-260).
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-debezium/src/main/java/com/ververica/cdc/debezium/event/DebeziumEventDeserializationSchema.java:165]: 👍🏻 for the detailed note

**Review by leonardBang — COMMENTED** [review]
Thanks @Jiabao-Sun for the great work, I left some comments

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/test/resources/ddl/column_type_test_mysql8.sql [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    decimal_c            DECIMAL(8, 4),
+    decimal_un_c         DECIMAL(8, 4) UNSIGNED,
+    decimal_un_z_c       DECIMAL(8, 4) UNSIGNED ZEROFILL,
+    numeric_c            NUMERIC(6, 0),
+    big_decimal_c        DECIMAL(65, 1),
+    bit1_c               BIT,
+    tiny1_c              TINYINT(1),
+    boolean_c            BOOLEAN,
+    PRIMARY KEY (id)
+) DEFAULT CHARSET=utf8;
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/test/resources/ddl/column_type_test_mysql8.sql:55]: Could we refer[1] to cover more types?
[1] https://github.com/ververica/flink-cdc-connectors/blob/4843ab2432fe77d3a019f8002ec765cf78e42a88/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/test/resources/ddl/column_type_test.sql#L55

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-debezium/src/main/java/com/ververica/cdc/debezium/event/DebeziumEventDeserializationSchema.java [hunk 13]
```diff
@@ -0,0 +1,537 @@
...
+            case INTEGER:
+                return convertToInt();
+            case BIGINT:
+                return convertToLong();
+            case DATE:
+                return convertToDate();
+            case TIME_WITHOUT_TIME_ZONE:
+                return convertToTime();
+            case TIMESTAMP_WITHOUT_TIME_ZONE:
+                return convertToTimestamp(serverTimeZone);
```
Comment (leonardBang) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-debezium/src/main/java/com/ververica/cdc/debezium/event/DebeziumEventDeserializationSchema.java:206]: After we support convert mysql timestamp to our timestamp_ltz, we do not need serverTimezone anymore for mysql datetime ?

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/utils/MySqlTypeUtils.java [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+            case DATE:
+                return DataTypes.DATE();
+            case DATETIME:
+                return column.length() >= 0
+                        ? DataTypes.TIMESTAMP(column.length())
+                        : DataTypes.TIMESTAMP();
+            case TIMESTAMP:
+                return column.length() >= 0
+                        ? DataTypes.TIMESTAMP_LTZ(column.length())
+                        : DataTypes.TIMESTAMP_LTZ();
```
Comment (lvyanquan) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-mysql/src/main/java/com/ververica/cdc/connectors/mysql/utils/MySqlTypeUtils.java:197]: DataTypes.TIMESTAMP() will be of TimestampType.DEFAULT_PRECISION, which is 6, we should set to 0 to be consistent with DebeziumSchemaDataTypeInference#infer method.

**Review by lvyanquan — CHANGES_REQUESTED** [review]

**Review by ruanhang1993 — APPROVED** [review]
LGTM.

**Pull request merged**